### PR TITLE
[TE] add generic database-backed configuration store

### PR DESCRIPTION
**Pull request by apucher**
* database-backed key-value store with namespaces for configuration properties (Jackson serializable)

    * strings

    * lists

    * maps

    * nesting

* config bean and DAO

* ConfigNamespace for safe put/get/delete access

* ConfigResource endpoint

* unit and integration tests

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  @POST
+  @Path("/{namespace}/{name}")
+  public void put(
+      @PathParam("namespace") String namespace,
+      @PathParam("name") String name,
+      String payload) throws IOException {
+
+    ConfigNamespace cn = makeNamespace(namespace);
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java]: what if the namespace does not exist. Are you going to create a new one?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    // determine suitable representation
+
+    // map
+    try {
+      cn.put(name, new ObjectMapper().readValue(payload, new TypeReference<Map<String, ?>>() {}));
+      LOG.info("Storing MAP '{}':'{}' = '{}'", namespace, name, payload);
+      return;
+    } catch (Exception ignore) {
+      // left blank
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java]: should we return some message here?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  @GET
+  @Path("/{namespace}/")
+  public Map<String, Object> list(
+      @PathParam("namespace") String namespace) {
+    return makeNamespace(namespace).getAll();
+  }
+
+  @DELETE
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java]: What's the corresponding operation for namespace, is it fixed? This related to previous comments.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/bao/ConfigManager.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.linkedin.thirdeye.datalayer.bao;
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/bao/ConfigManager.java]: should we call it DatabaseConfigManager?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java (hunk 1 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java]: It's transparent.
https://stackoverflow.com/questions/19279628/what-is-a-good-way-to-manage-keys-in-a-key-value-store

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java (hunk 2 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java]: please re-read comments and documentation

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java (hunk 3 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/ConfigResource.java]: see above

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/bao/ConfigManager.java (hunk 4 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/bao/ConfigManager.java]: Either Config or Settings

**Comment by apucher** [comment]
@bl44 this change is required to support re-configuration of thirdeye components at runtime

**Pull request merged**